migrations/20241105164503_2.sql

```diff
@@ -0,0 +1,40 @@
+-- Force anything running this migration to use the right search path.
+set local search_path to underway;
+
+-- Remove attempt-specific columns from `task`
+alter table underway.task
+drop column if exists retry_count,
+drop column if exists initial_interval_ms,
+drop column if exists max_interval_ms,
+drop column if exists backoff_coefficient,
+drop column if exists error_message,
+drop column if exists started_at,
+drop column if exists succeeded_at,
+drop column if exists last_failed_at;
+
+alter table underway.task
+add column if not exists retry_policy jsonb not null default '{"max_attempts": 5, "initial_interval_ms": 1000, "max_interval_ms": 60000, "backoff_coefficient": 2.0}';
+
+alter table underway.task
+add column if not exists completed_at timestamp with time zone;
+
+alter table underway.task
+add column if not exists last_attempt_at timestamp with time zone;
+
+create table underway.task_attempt (
+    task_id         uuid not null,
+    task_queue_name text not null,
+    attempt_number  integer not null,
+
+    -- Task state.
+    state           underway.task_state not null default 'pending',
```

Comment:
small one, dont think it can be `pending`, I would expect that attemp inserted once it happen, so it is failure or success

Reply:
I agree "pending" is probably not right. The attempt is created upon dequeue (this is when we actually start the attempt). So this should likely be "in_progress" as a reasonable default.

Reply:
in my implementation i didnt bother to have `in_progress`, I've inserted only finished attempts (failed or succeded)

Reply:
I thought about that too but I think I prefer it this way for instrumentation purposes: by creating the row when we dequeue we'll be able to derive more accurate metrics about task attempts (i.e. it'll incorporate the time it took to execute the task).

It's also a little easier reason about derived properties knowing that the attempt row will always exist once it's been dequeued.

Reply:
I am under impression that because of tx nature - no one will see this attempt in `in_progress` state.
Please correct me if I am wrong

Reply:
The queue itself doesn't require a transaction (e.g. `dequeue` takes a `PgConnection`). The provided worker implementation does use a transaction, so in that case it should be true.

Holding a transaction open for the duration of task execution is probably okay most of the time but I want to retain the possibility for extending the crate with workers that might not work that way.

Even if we don't end up providing that right away, creating attempt rows in tandem with a dequeue is beneficial in a least a couple of ways:

1. Encapsulating the state transition in the queue itself is an arguably more sound design (it means a caller doesn't have to remember to e.g. create an attempt row later, or even know about attempt rows at all, and keeps the overall design more coherent in my view) and,
2. Because attempt rows are created during the dequeue, their timing metrics will be better aligned with the moment a task was dequeued--this could be done manually, but I'm not really sure what we gain by doing that.

If later we're positive that we'll never support out-of-transaction workers I think we could still make a change that constrains the states and removes "in progress" so this isn't necessarily set in stone.